bansa/src/main/java/com/brianegan/bansa/BansaStore.java

```diff
@@ -0,0 +1,68 @@
+package com.brianegan.bansa;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class BansaStore<S, A> implements Store<S, A> {
+    private S currentState;
+    private final Reducer<S, A> reducer;
+    private final List<Subscriber> subscribers = new ArrayList<>();
+    private final List<NextDispatcher<A>> dispatchers = new ArrayList<>();
+    private final Middleware<S, A> dispatcher = new Middleware<S, A>() {
+        @Override
+        public void dispatch(Store<S, A> store, A action, NextDispatcher<A> next) {
+            synchronized (this) {
+                currentState = reducer.reduce(store.getState(), action);
+            }
+            for (int i = 0; i < subscribers.size(); i++) {
+                subscribers.get(i).onStateChange();
+            }
+        }
+    };
+
+    @SafeVarargs
+    public BansaStore(S initialState, Reducer<S, A> reducer, Middleware<S, A>... middlewares) {
+        this.reducer = reducer;
+        this.currentState = initialState;
+        final Store<S, A> store = this;
+
+        dispatchers.add(new NextDispatcher<A>() {
+            public void dispatch(A action) {
+                dispatcher.dispatch(store, action, null);
+            }
+        });
+
+        for (int i = middlewares.length - 1; i >= 0; i--) {
+            final Middleware<S, A> middleware = middlewares[i];
+            final NextDispatcher<A> next = dispatchers.get(0);
+            dispatchers.add(0, new NextDispatcher<A>() {
```

Comment:
Yah, good point about the array shifting that happens adding at the beginning, that's actually what I was thinking about when I read your point! I just wasn't sure of that vs the cost of looping linearly through the array (as `Collections.reverse` does).

But, as you said, `LinkedList` would have constant time adding to the beginning, so I'll give that a shot!

Thanks so much for the good discussions! Love open source :)